Generate serialization for WCLayerUpdateInfo.

nishajain61 opened the pull request:
#### 107335f965fa12928143b48c179274eb066a1691
<pre>
Generate serialization for WCLayerUpdateInfo.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267600">https://bugs.webkit.org/show_bug.cgi?id=267600</a>
<a href="https://rdar.apple.com/121069952">rdar://121069952</a>

Reviewed by Chris Dumez.

Generate serialization for WCLayerUpdateInfo.

* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
(WebKit::WCLayerUpdateInfo::encode const): Deleted.
(WebKit::WCLayerUpdateInfo::decode): Deleted.
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:

Canonical link: <a href="https://commits.webkit.org/273862@main">https://commits.webkit.org/273862@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8484ae5ab610e13a5d832b1b4b121178bef98d78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36905 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39529 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18356 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12956 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37467 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32589 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/11672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40781 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33459 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33226 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11970 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9775 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35709 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4785 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12881 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/WebKit/DerivedSources.make:
```diff
@@ -761,6 +761,7 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	WebProcess/UserContent/InjectUserScriptImmediately.serialization.in \
 	WebProcess/WebCoreSupport/WebSpeechSynthesisVoice.serialization.in \
 	WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemoteProperties.serialization.in \
+	WebProcess/WebPage/wc/WCUpdateInfo.serialization.in \
```
I don't think you need to add this here since Apple code is not using this.

donny-dont reviewed (approved):
Looks correct assuming Windows build works. Thanks for doing this!

sheeparegreat reviewed (changes requested):
I believe you also need to update `Source/WebKit/DerivedSources-input.xcfilelist` too?

donny-dont commented on Source/WebKit/CMakeLists.txt:
```diff
@@ -588,6 +588,8 @@ set(WebKit_SERIALIZATION_IN_FILES
 
     WebProcess/WebCoreSupport/WebSpeechSynthesisVoice.serialization.in
+
+    WebProcess/WebPage/wc/WCUpdateInfo.serialization.in
```
Whoops caught this. It should just be in `PlatformWin.cmake`. It is not used by other ports

donny-dont replied:
Actually just delete this since you're appending to the `serialization.in` file and its already present in the `PlatformWin.cmake` file.

sheeparegreat commented on Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:
```diff
@@ -39,4 +39,65 @@ struct WebKit::WCUpdateInfo {
     Vector<WebKit::WCLayerUpdateInfo> changedLayers;
 }
 
+[Nested,OptionSet] enum class WebKit::WCLayerChange : uint8_t {
```
In the code it's defined as: `enum class WCLayerChange : uint32_t {`. This needs to be consistent

donny-dont replied:
This should be uint32_t

sheeparegreat commented on Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:
```diff
@@ -39,4 +39,65 @@ struct WebKit::WCUpdateInfo {
+};
+
+[CustomHeader] struct WebKit::WCLayerUpdateInfo {
+    [Validator='id'] WebCore::PlatformLayerIdentifier id;
```
I think we can remove the validator here as it will return std::nullopt if it does not have enough data
```
    auto id = decoder.decode<WebCore::PlatformLayerIdentifier>();
    if (UNLIKELY(!decoder.isValid()))
        return std::nullopt;
```
We are not constraining id to a value so the validator isn't adding value

donny-dont commented on Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
```diff
@@ -45,7 +45,7 @@ struct WCTileUpdate {
 
-enum class WCLayerChange : uint32_t {
+enum class WCLayerChange : uint8_t {
     Children                = 1 <<  0,
```
Revert this change and mark the serialization files definition as uint32_t

This is why compilation is failing

sheeparegreat commented on Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:
```diff
@@ -39,4 +39,65 @@ struct WebKit::WCUpdateInfo {
+
+[CustomHeader] struct WebKit::WCLayerUpdateInfo {
+    [Validator='id'] WebCore::PlatformLayerIdentifier id;
+    [OptionalTupleBits] OptionSet<WCLayerChange> changes;
```
WebKit::WCLayerChange

cdumez reviewed (approved):
LGTM if the bots are happy.

nishajain61 committed:
Generate serialization for WCLayerUpdateInfo.
https://bugs.webkit.org/show_bug.cgi?id=267600
rdar://121069952

Reviewed by Chris Dumez.

Generate serialization for WCLayerUpdateInfo.

* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
(WebKit::WCLayerUpdateInfo::encode const): Deleted.
(WebKit::WCLayerUpdateInfo::decode): Deleted.
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.serialization.in:

Canonical link: https://commits.webkit.org/273862@main
